Add plug&play Dockerfile, fix imports, unpin requirements

ddelange opened the pull request:
This PR is basically a trial&error walk through your Installation wiki, accumulating all those steps into a ready-to-use Dockerfile (uses your db.sqlite3 file by default, although the ADMIN_PASSWORD=password `ENV` statement obviously won't work out of the box with your db). Please do squash if you like the PR, to get rid of all the dev commits 😅  the diff without whitespaces is about half the size.

- Add plug&play Dockerfile, the django app now respects DEBUG, DATABASE_URL and ADMIN_ env vars (in case the db doesn't have an admin yet).
- Fix relative and wildcard imports, also for Adaptive-NER methods that were previously still commented out. I renamed `from alpaca_serving.client import AlpacaClient` to `from alpaca_serving_client.client import AlpacaClient` due to python package namespace conflict with the `alpaca_serving` package when installed in the same env (the Dockerfile).
- Unpin requirements. With pips new resolver (20.3), lower pins are perfectly fine. The Dockerfile boots fine with django3, latest pytorch etc.
- Switch from glove to bert in `SequenceTaggingModel` to avoid (hardcoded) glove embedding asset download error (ref https://github.com/flairNLP/flair/issues/1831)

ddelange committed:
Add initial Dockerfile, plus a copy from doccano/doccano

ddelange committed:
Rewrite with pip install

ddelange committed:
Edit

ddelange committed:
Edit

ddelange committed:
Edit

ddelange committed:
Add gcc

ddelange committed:
pyodbc error

ddelange committed:
Efficient pip install

ddelange committed:
Efficiency

ddelange committed:
Efficiency

ddelange committed:
Editable install

ddelange committed:
Regular joblib

ddelange committed:
Edit

ddelange committed:
Rename alpaca_serving_client

ddelange committed:
Typo

ddelange committed:
Add entrypoint

ddelange committed:
Add entrypoint

ddelange committed:
Add entrypoint

ddelange committed:
Automate superuser creation

ddelange committed:
Delete doccano Dockerfile

ddelange committed:
Quotes

ddelange committed:
Add startup.sh

ddelange committed:
Unescape

ddelange committed:
Adopt doccano run.sh

ddelange committed:
Typo

ddelange committed:
Typo

ddelange committed:
Typo

ddelange committed:
Loosen requirements

ddelange committed:
Typo

ddelange committed:
Delete db

ddelange committed:
Fix createsuperuser

ddelange committed:
Fix missing import

ddelange committed:
Fix missing import

ddelange committed:
Remove homepage redirect, go to projects instead

ddelange committed:
Startup typo

ddelange committed:
Add more ENV entries

ddelange committed:
Revert "Delete db"

This reverts commit fa32283b1a851c0cad76b513ad1a198d5295e2f9.

ddelange committed:
Avoid star import

ddelange committed:
Don't init db

ddelange committed:
Revert

ddelange committed:
Don't create admin

ddelange committed:
Use django-environ to read DATABASE_URL

ddelange committed:
Correct default db in Dockerfile

ddelange committed:
Fix all wildcard and relative imports using removestar package

ddelange committed:
Switch to bert embeddings

ddelange committed:
Typo

ddelange committed:
Dev leftover

ddelange commented on Dockerfile:
```diff
@@ -0,0 +1,66 @@
+FROM python:3.6-slim-buster as final
```
as alternative base image, I considered [`pytorch/pytorch:1.6.0-cuda10.1-cudnn7-runtime`](https://hub.docker.com/r/pytorch/pytorch/tags?page=1&ordering=last_updated), which runs py3.7 and will optionally pick up GPUs with a `docker run --gpus all ...` if the host runner has the NVIDIA Container Toolkit [installed](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#setting-up-nvidia-container-toolkit).

Rahul-Khanna commented:
hi @ddelange, I would be open to doing a call to understand all these changes. I'm decently familiar with this system, and am building another one based off this project (https://github.com/INK-USC/LEAN-LIFE), so would love to dockerize that project. Let me know if that works for you. I also can merge this request in once I understand what's going on.

Rahul-Khanna commented:
thanks for doing this by the way!

ddelange commented:
Assuming you're in CA: good morning @Rahul-Khanna! 

You can reach me at t.me/Rhiniodon (my tz is GMT+1 btw), I should be able to do a quick Zoom or so spontaneously.

ddelange committed:
Embedding error fix

File "/alpaca/alpaca_server/alpaca_model/pytorchAPI/wrapper.py", line 108, in online_word_build
embeddings = get_Bert_embeddings(self.p._word_vocab.vocab, BERT_EMBEDDING_SIZE)
File "/alpaca/alpaca_server/alpaca_model/pytorchAPI/utils.py", line 188, in get_Bert_embeddings
_embeddings[vocab[token.text]] = token.embedding
KeyError: '<'

ddelange committed:
Import typos

ddelange committed:
Echo typo

ddelange committed:
Try Elmo

ddelange committed:
Missing allennlp module

ddelange committed:
Missing allennlp module -- try conda

ddelange committed:
Add DistilBERT embeddings

ddelange committed:
Remove unused import

ddelange committed:
Revert

ddelange commented:
So with eee2c98, `alpaca-serving` started successfully responding to django. However, it seems that there are some front-end issues with the current codebase (probably somewhere [here](https://github.com/ddelange/AlpacaTag/blob/ddbf3caaa566ee11f297067025c2a1d115d96b96/annotation/AlpacaTag/server/static/js/sequence_labeling.js#L43-L64)), especially with the annotation itself (some offset of around 20 chars versus what is selected). My hope here is that somehow merging a (way) more recent version of doccano into the codebase would fix it. Do you have time/ideas how to go about that?

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/14880945/102473704-0f203200-4058-11eb-9368-e35af6c4d0a1.gif)

ddelange committed:
Hardcoded 20char subtration

https://github.com/INK-USC/AlpacaTag/pull/12#issuecomment-747341763

ddelange committed:
Revert subtraction, remove 20 hardcoded spaces

ddelange committed:
Put template on one line

ddelange committed:
Remove bundle folder (from webpack)

ddelange committed:
Add static/bundle/ to gitignore

ddelange committed:
Add .dockerignore

ddelange committed:
Revert trials

ddelange commented:
So looking at the `div`, there are actually exactly 20 leading spaces:

![image](https://user-images.githubusercontent.com/14880945/102597329-c8dfd700-411a-11eb-9514-6b25e7482e85.png)


I tried finding them, thought it was these 20 spaces (but it wasnt) https://github.com/ddelange/AlpacaTag/blob/75f9723765ea2ba4deb296ef7cd0e7ef42278c7e/annotation/AlpacaTag/server/static/js/sequence_labeling.js#L25

and checked out the corresponding html but didn't find the cause.
https://github.com/ddelange/AlpacaTag/blob/75f9723765ea2ba4deb296ef7cd0e7ef42278c7e/annotation/AlpacaTag/server/templates/annotation/sequence_labeling.html#L29

ddelange committed:
Remove another 20 spaces inside div, before span

ddelange committed:
Add .env to gitignore

ddelange committed:
Delete db.sqlite3

ddelange committed:
Add back db.sqlite3 to gitignore

ddelange committed:
Make docker image smaller

ddelange committed:
Run npm audit fix

ddelange committed:
Add comments to Dockerfile

ddelange committed:
Revert "Add static/bundle/ to gitignore"

This reverts commit b2f1a59b5b7f6921c9a1f45ae0b531e25c094f5c.

ddelange committed:
Revert "Remove bundle folder (from webpack)"

This reverts commit 007228230512b340b811948d8b648d86f6c5edc9.

ddelange committed:
Remove 20 spaces from /bundle/sequence_labeling.js

ddelange committed:
Revert "Run npm audit fix"

This reverts commit 602cd57805765e738b031b9c44f8e16c92462d15.

ddelange commented:
Alright. Tagging is working now! I wanted webpack (`npm run build`) to build the `bundle` folder on the fly (instead of having it in git), but webpack seems to not build the bundle folder so I resorted to removing the 20 spaces manually from the bundle js (a03a001) as well as from the source js (9893f53). This did the trick :)

danny911kr commented:
Hi, @ddelange Thank you so much for doing this.
I just get involved in the review of your huge efforts!
I'll get back to you as soon as possible.
